src/dataframe/dataframe.jl

```diff
@@ -394,65 +393,78 @@ end
+
+As a special case, `df[ncol(df)+1] = v` is allowed and creates a new column at the end
+of `df`.
+"""
+Base.setindex!(df::DataFrame, v, col_ind)
+
+"""
     setindex!(df::DataFrame, v, col_inds::AbstractVector})
 
-`col_inds` can be either vector of `Symbol`, `Integer`, `Union{Bool, Missing}` or `:`.
+Assign `v` to columns `col_inds` of data frame `df`.
 
-If `v` is not a `DataFrame` then assignment rules follow single column `setindex!`.
+`col_inds` can be either `:`, a vector of `Symbol` or `Integer`, or a vector of `Bool`
+of the same length as the number of columns in `df` (logical indexing).
 
-If `v` is a `DataFrame` then assignment to `col_inds` take columns from `v` sequentially
-disregarding their names.  Number of rows of `df` and `v` must be identical in this case.
-Additionally number of columns in `v` must be equal to length of `col_inds`.
+If `v` is not a data frame then assignment rules follow single-column `setindex!` .
 
-You can create new columns in `df` this way and the rules are the same as for `setindex!`
-on a single column.
+If `v` is a data frame then each column in `v` is assigned sequentially to the column
+at the corresponding position in `col_inds`, disregarding column names.
+`df` and `v` must have the same number of rows, and the number of columns in `v`
+must be equal to the length of `col_inds`.
 
-If `col_inds<:AbstractVector{Union{Bool, Missing}}` then it must only contain `Bool` values
-and its length must be equal to number of columns in `df`.
+New columns can be added to `df` and the rules are the same as for `setindex!` on a single column.
 
-Assignment of columns `col_inds` to `df` is performed sequentially which matters if
+Assignment of columns `col_inds` to `df` is performed sequentially, which matters if
 `col_inds` contains duplicates or if one wants to create multiple new columns in `df` using
-`Integer` column index.
-
-
-    setindex!(df::DataFrame, v, row_ind, col_ind)
-
-The rules are the same as for inserting whole columns with the following additional convention:
-* you cannot create new columns when passing row index (except if it is `:` see last rule below)
-* row indexing follows standard indexing rules in Base
-* if `v` is `AbstractVector` its length must be equal to number of rows selected by `row_ind`
-  and its elements must be convertible to a type of elements in the column assigned to;
-  otherwise `v` is treated as scalar and broadcasted to all entries of a target column except
-  when it is a `DataFrame` in which case the same rules as for column assignment of `DataFrame`
-  apply with an additional restriction that `v` must have the same number of rows as selected
-  by `row_ind`;
-* if `row_ind` is `:` then it behaves exactly like `setindex!` without row selection except
-  when `v` is a `DataFrame` in which case `df[:]` and `df[:,:]` performs a copy of `v` to `df`
-
-In all above rules if `DataFrame` is mentioned on right hand side of an assignment it may not
```

Comment:
I've removed this paragraph since this sounds like a bug to me, isn't it? If so we should file an issue.